[MM-45946]capture hyperlinks when pasting

EricssonLiu opened the pull request:
#### Summary
This ticket adds functionality to have markdown formatted links when a hyperlink is pasted into the post or comment text editors.

Retuned pasteHandlers for creating and editing posts and comments to use turndown when it encounters specifically tables and/or links. Headerless tables will continue to use old in-house formatter due to lack of plugin support. 

TODO: 
- [x] add a few unit tests

~~Google Docs/Microsoft Word paste handling~~ (followup PR)

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/20704

https://mattermost.atlassian.net/browse/MM-45946

#### Release Note

```release-note
Added markdown formatting for hyperlinks when pasted into the text editor.
```

michelengelen commented on components/advanced_create_post/advanced_create_post.test.jsx:
```diff
@@ -1239,6 +1269,36 @@ describe('components/advanced_create_post', () => {
+        const markdownLink = '[link text](https://test.domain)';
+
+        wrapper.instance().pasteHandler(event);
+        expect(wrapper.state('message')).toBe(markdownLink);
```
could you add a check for the change in a draft here as well please. Not really needed, but better to be more careful about these to catch more errors

EricssonLiu replied:
as far as I see I think the pasteHandler in `advancedCreatePost` doesn't actually update the draft, only does it in comment's pasteHandler

michelengelen replied:
yes, it does it [here](https://github.com/mattermost/mattermost-webapp/blob/5f378028908969463c57af109755b2b7de30cc5f/components/advanced_create_post/advanced_create_post.tsx#L362)

EricssonLiu replied:
Took a further look into testing, tried to set something up but draft didn't seem to update, did some actual manual testing and looks like it doesn't save draft when pasting tables/links. (both before this change (when pasting tables) and now (when any paste goes through the handler)). Should be a relatively simple addition to fix but should this probably be another ticket? @esethna

michelengelen commented on utils/paste.tsx:
```diff
@@ -1,5 +1,8 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
+import TurndownService from 'turndown';
+import {tables} from 'turndown-plugin-gfm';
```
just use this, since `turndown-plugin-gfm` does not support esm:

```suggestion
const tables = require('turndown-plugin-gfm').tables;
```

michelengelen replied:
It would be great if you could add a small comment on top of this line explaining why `require` is used instead of `import`

EricssonLiu replied:
Not sure what the differences are between these, left this for now

michelengelen replied:
@hmhealey I just wanted to keep the amount of new file creations at a minimum... 

@EricssonLiu it is just how the module is resolved ... esm is the easiest, since you can just use `import`. `require` is the 'old' way of doing it! :D

michelengelen commented on types/external/turndown.d.ts:
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+declare module 'turndown-plugin-gfm';
```
no need to redeclare.
I added a comment below to the import part of the plugin

michelengelen commented on utils/paste.tsx:
```diff
@@ -32,6 +35,15 @@ export function getTable(clipboardData: DataTransfer): HTMLTableElement | null {
     return table;
 }
 
+export function getHasLinks(clipboardData: DataTransfer): boolean {
```
naming is misleading. better use: `hasHtmlLink`

michelengelen commented on utils/paste.tsx:
```diff
@@ -32,6 +35,15 @@ export function getTable(clipboardData: DataTransfer): HTMLTableElement | null {
 }
 
+export function getHasLinks(clipboardData: DataTransfer): boolean {
+    if (Array.from(clipboardData.types).indexOf('text/html') === -1) {
+        return false;
+    }
+
+    const html = clipboardData.getData('text/html');
+    return (/<a/i).test(html);
+}
```
can be simplified a bit:

```
return Array.from(clipboardData.types).indexOf('text/html') > -1 && (/<a/i).test(clipboardData.getData('text/html'));

hmhealey replied:
If we're changing that, you can also change the `.indexOf('text/html') > -1` to `.includes('text/html)`.

michelengelen commented on utils/paste.tsx:
```diff
@@ -71,13 +87,30 @@ export function formatMarkdownTableMessage(table: HTMLTableElement, message?: st
     }).join('\n');
 
     const formattedTable = `${header}${body}\n`;
+    return formattedTable;
```
nit and no mistake from you, but this variable is unnecessary:

```suggestion
    return `${header}${body}\n`;
```

michelengelen commented on utils/paste.tsx:
```diff
@@ -71,13 +87,30 @@ export function formatMarkdownTableMessage(table: HTMLTableElement, message?: st
+export function formatMarkdownMessage(clipboardData: DataTransfer, message?: string, caretPosition?: number): string {
+    const html = clipboardData.getData('text/html');
+
+    const service = new TurndownService({emDelimiter: '*'});
```
we should probably create a service from a file that has all of our settings baked in already. This is out of scope for this PR, but could you add a line with a `TODO@michel` in it? I need to create an instance of it anyways for the WYSIWYG editor

michelengelen reviewed (changes requested):
a few comments from my side, but it looks really good otherwise

michelengelen reviewed (approved):
LGTM

esethna reviewed (changes requested):
@EricssonLiu @michelengelen this doesn't seem to be working with hyperlinks inside tables. ie I tried copying a few cells from google sheets with hyperlinks and the link is still lost in the paste

Is this something we can accommodate for?

EricssonLiu commented:
@esethna @michelengelen so I know why this is happening - when we just take cells from sheets, it's a headerless table and thus we have to defer to the old markdown table formatter instead of turndown, because for some reason when theres no headers the table plugin in `turndown-plugin-gfm` [returns the raw html instead of trying to format](https://github.com/mixmark-io/turndown-plugin-gfm/blob/master/src/tables.js) and it was specifically made as a change. I didn't account that most table entries would probably be headerless.

For Michel, do you think it would make sense if we're using turndown in wysiwyg as well to build our own turndown rules files? Then we could probably capture headerless tables as well that way, unless theres a good reason why they were discluded. Otherwise we could have the turndown service run on each cell in the table in our in-house table formatter.

EricssonLiu commented:
@michelengelen @esethna Changed to the fork of the table plugin, and added some filtering to make the google sheets pastes work. Also made a change to make single row tables paste better and having tables pasted between a single-line message format correctly. Since headerless tables are now covered, didn't need to use our old in-house table converter anymore and linter had me remove the functions around that. Let me know if we prefer to comment out the linter to preserve those functions.

EricssonLiu commented on utils/paste.tsx:
```diff
@@ -47,37 +54,41 @@
+    if (isGoogleSheetsPaste(clipboardData)) {
+        // Removes an html comment that is kept
+        const firstLineRegex = /<.*>[\r\n]+/g;
+        markdownFormattedMessage = markdownFormattedMessage.replace(firstLineRegex, '');
```
uhhh.. this shouldn't be an actual concern, right? I don't really see where it could add any vulnerabilities?

michelengelen commented:
> @michelengelen @esethna Changed to the fork of the table plugin, and added some filtering to make the google sheets pastes work. Also made a change to make single row tables paste better and having tables pasted between a single-line message format correctly. Since headerless tables are now covered, didn't need to use our old in-house table converter anymore and linter had me remove the functions around that. Let me know if we prefer to comment out the linter to preserve those functions.

IMO since we replace all of the usages we can just remove it. If this adds a regeression this will get caught by the tests. Or, with the very worst situation, we can just roll back the changes. 👍🏼

EricssonLiu commented:
> @EricssonLiu thanks for the troubleshooting here. There seems to be a regression with excel tables now:

@esethna Thanks for the catch, I've changed the filtering to be source-agnostic and so it should hopefully catch from all paste sources now, instead of only working for google sheets (and added a test).

esethna commented:
@EricssonLiu, hmm when i try from excel the link is lost now

![image](https://user-images.githubusercontent.com/14333569/196272337-da68e838-b133-495c-a2f0-95c7b3c02b5a.png)

esethna reviewed (approved):
Thanks @EricssonLiu!

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/251852

esethna commented:
@EricssonLiu this is an incredibly high impact improvement pasting, great work with it!! Let's get this reviewed and merged!

AshishDhama commented:
@jgilliam17 its ready for QA

jgilliam17 reviewed (approved):
Thanks @EricssonLiu 
Tested, looks good to merge.
Verified pasting hyperlinks is retaining the markdown.
[E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/251852-afbbb5e-onprem-ent-webapp-pr-11299-mm-ee-test:afbbb5e/mochawesome.html) - no PR related failures.
